lib/TableGen/Operations.cpp

```
@@ -363,6 +363,47 @@ bool Operation::parse(raw_ostream &errs, GenDialectsContext *context,
   for (RecordTy *traitRec : record->getValueAsListOfDefs("traits"))
     op->traits.push_back(context->getTrait(traitRec));
 
+
+  const RecordVal *insVal = record->getValue("arguments");
+  std::unordered_map<std::string, unsigned> nameToIndexMap;
+  if (const DagInit *DI = dyn_cast<DagInit>(insVal->getValue())){
```

It is an error for the `arguments` field to not be of dag type, and we should err on the side of failing loudly on unexpected data. Otherwise, users may end up specifying their operations incorrectly without noticing it.

Just use `getValueAsDag` here, as is done elsewhere. And the same applies below to `results` and analogously to `value_traits`.

---

Agree for the `arguments` field. Changed it to use `getValueAsDag`. However, `results` might be empty when working with `OpClass`es. For the `value_traits`, I added the `fatal_error_report` in case there is something other than dag inside of `value_traits` list.

---

For the `results` case, you can still assume that `results` is a dag. That should even be ensured by the type system of the TableGen frontend. The only check you gneed is for `getNumArgs`.

---

Sorry, I should've expressed myself better. It would be correct to assume `results` is a dag, but only when there is a `results` field. In case of `OpClass` there is no `results` field. When field is not present, `getValueAsDag` returns fatal error. Seems to me things should stay as is for the `results` case.